ethcore/src/miner/stratum.rs

```diff
@@ -0,0 +1,190 @@
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use ethcore_stratum::{JobDispatcher, RemoteWorkHandler, PushWorkHandler};
+use std::sync::{Arc, Weak};
+use std::sync::atomic::Ordering;
+use std::thread;
+use nanoipc;
+use util::{H256, U256, FixedHash, H64};
+use ethereum::ethash::Ethash;
+use ethash::SeedHashCompute;
+use util::{Mutex, RwLock};
+use miner::{Miner, MinerService};
+use client::Client;
+use block::IsBlock;
+use std::str::FromStr;
+use util::rlp::encode;
+
+/// Job dispatcher for stratum service
+pub struct StratumJobDispatcher {
+	last_work: RwLock<Option<(H256, U256, u64)>>,
+	seed_compute: Mutex<SeedHashCompute>,
+	client: Weak<Client>,
+	miner: Weak<Miner>,
+}
+
+impl JobDispatcher for StratumJobDispatcher {
+	fn initial(&self) -> Option<String> {
+		// initial payload may contain additional data, not in this case
+		self.job()
+	}
+
+	fn job(&self) -> Option<String> {
+		{
+			let last_work = self.last_work.write();
+			if let Some((pow_hash, difficulty, number)) = *last_work {
+				return Some(self.payload(pow_hash, difficulty, number));
+			}
+		}
+
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
+
+		if let Some((pow_hash, difficulty, number)) = miner.map_sealing_work(&*client, |b| {
+			let pow_hash = b.hash();
+			let number = b.block().header().number();
+			let difficulty = b.block().header().difficulty();
+
+			(pow_hash, *difficulty, number)
+		}) {
+			*self.last_work.write() = Some((pow_hash, difficulty, number));
+			Some(self.payload(pow_hash, difficulty, number))
+		} else { None }
+	}
+
+	fn submit(&self, payload: Vec<String>) {
+		if payload.len() != 3 {
+			warn!(target: "stratum", "submit_work: invalid work submit request({:?})", payload);
+		}
+
+		let nonce = match H64::from_str(&payload[0]) {
+			Ok(nonce) => nonce,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid nonce ({:?})", e);
+				return;
+			}
+		};
+
+
+		let pow_hash = match H256::from_str(&payload[1]) {
+			Ok(pow_hash) => pow_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid hash ({:?})", e);
+				return;
+			}
+		};
+
+		let mix_hash = match H256::from_str(&payload[2]) {
+			Ok(mix_hash) => mix_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid mix-hash ({:?})", e);
+				return;
+			}
+		};
+
+		trace!(target: "stratum", "submit_work: Decoded: nonce={}, pow_hash={}, mix_hash={}", nonce, pow_hash, mix_hash);
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
+
+		let seal = vec![encode(&mix_hash).to_vec(), encode(&nonce).to_vec()];
+		if let Err(e) = miner.submit_seal(&*client, pow_hash, seal) {
+			warn!(target: "stratum", "submit_work error: {:?}", e);
+		};
+	}
+}
+
+impl StratumJobDispatcher {
+	/// New stratum job dispatcher given the miner and client
+	fn new(miner: &Arc<Miner>, client: &Arc<Client>) -> StratumJobDispatcher {
```

Comment:
style: I think functions like this which take an `Arc` but actually downgrade are pretty misleading -- better to call a spade a spade IMO and make it clear from the interface.